lib/src/package_analysis/common.dart

```diff
@@ -1,12 +1,204 @@
+/// [destination] and fetch its dependencies, optionally deleting the contents
+/// of [destination] first if [wipeTarget] is set to true.
+///
+/// The package will be downloaded using `downloadPackage` from
+/// `package:pana/src/download_utils.dart`.
+Future<void> fetchPackageAndDependencies({
+  required String name,
+  required String version,
+  required String destination,
+  required bool wipeTarget,
+}) async {
+  // delete the target directory, if it exists and wipe is enabled
+  if (wipeTarget && await Directory(destination).exists()) {
+    await Directory(destination).delete(recursive: true);
+  }
+
+  // download the package and fetch dependencies
+  await downloadPackage(name, version, destination: destination);
+  await fetchDependencies(destination);
+  return;
+}
+
+/// Download version [version] of the package [name] to the directory
+/// [destination] and fetch its dependencies, optionally deleting the contents
+/// of [destination] first if [wipeTarget] is set to true.
+///
+/// A pointer package will be created at [destination] with a single dependency,
+/// pinning the dependency [name] to version [version].
+Future<void> fetchPackageWithPointer({
+  required String name,
+  required String version,
+  required String destination,
+  required bool wipeTarget,
+}) async {
+  // delete the target directory, if it exists and wipe is enabled
+  if (wipeTarget && await Directory(destination).exists()) {
+    await Directory(destination).delete(recursive: true);
+  }
+
+  // construct pubspec for the pointer package
+  final pointerPubspec = <String, dynamic>{
+    'name': 'dummy_package',
+    'environment': {
+      'sdk': '>=2.12.0 <3.0.0',
+    },
+    'dependencies': {},
+  };
+  pointerPubspec['dependencies'][name] = version;
+
+  // if pubspec.yaml exists, delete it and create an empty file
+  final pubspecFile = File(path.join(destination, 'pubspec.yaml'));
+  if (await pubspecFile.exists()) {
+    await pubspecFile.delete();
+  }
+  await pubspecFile.create(recursive: true);
+
+  // write pubspec to disk
+  await pubspecFile.writeAsString(json.encode(pointerPubspec));
+
+  // fetch dependencies (the only non-transitive dependency is [name])
+  await fetchDependencies(destination);
+}
+
+/// Fetches dependencies in [destination].
+Future<void> fetchDependencies(String destination) async {
+  final result = await Process.run(
+    Platform.resolvedExecutable,
+    ['pub', 'get'],
+    workingDirectory: destination,
+  );
+  if (result.exitCode != 0) {
+    throw ProcessException(
+      Platform.resolvedExecutable,
+      ['pub', 'get'],
+      result.stderr.toString(),
+      result.exitCode,
+    );
+  }
+}
+
+/// Given the location of a package and the name of one of its dependencies,
+/// return the path of the dependency, or null if it cannot be resolved.
+/// Ensure that the dependencies of the target package are fetched.
+Future<String?> getDependencyDirectory(
```

Comment:
```suggestion
Future<String?> findPackageDirectory(
```

Or maybe just make an extension method on `packageAnalysisContext`? or even a method on `packageAnalysisContext` called `packageAnalysisContext.packageDirectory(String packageName)` or even just `packageAnalysisContext.packageFolder(String packageName)`.

Maybe, I also suggest that we make it non-nullable?
If the package doesn't exist then something is wrong, right? Like we failed to run `dart pub get` or we have a bug in our analysis code because we are looking for a package that wasn't a dependency or transitive dependency?
So it should never return null, correct? unless there is a bug in our code.

Thus, I suggest that we simply make it non-nullable and if we fail to resolve it (in cases where `dependencyFilePath == null`), we just throw  an `StateError` saying that the referenced package doesn't exist.
Because it's always an error if we try to analyze a package that we don't depend on..